trunk/source/BuchiAutomizer/src/de/uni_freiburg/informatik/ultimate/plugins/generator/buchiautomizer/cegar/BuchiCegarLoopFactory.java

```diff
@@ -0,0 +1,140 @@
+import de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction.witnesschecking.WitnessUtils.Property;
+import de.uni_freiburg.informatik.ultimate.witnessparser.graph.WitnessEdge;
+import de.uni_freiburg.informatik.ultimate.witnessparser.graph.WitnessNode;
+
+/**
+ *
+ * @author Frank Schüssele (schuessf@informatik.uni-freiburg.de)
+ *
+ * @param <L>
+ */
+public class BuchiCegarLoopFactory<L extends IIcfgTransition<?>> {
+	private final BuchiCegarLoopBenchmarkGenerator mCegarLoopBenchmark;
+	private final Class<L> mTransitionClazz;
+
+	public BuchiCegarLoopFactory(final Class<L> transitionClazz,
+			final BuchiCegarLoopBenchmarkGenerator benchmarkGenerator) {
+		mTransitionClazz = transitionClazz;
+		mCegarLoopBenchmark = benchmarkGenerator;
+	}
+
+	public AbstractBuchiCegarLoop<L, ?> constructCegarLoop(final IIcfg<?> icfg,
+			final RankVarConstructor rankVarConstructor, final PredicateFactory predicateFactory,
+			final TAPreferences taPrefs, final IUltimateServiceProvider services,
+			final INestedWordAutomaton<WitnessEdge, WitnessNode> witnessAutomaton) {
+		final PredicateFactoryRefinement stateFactoryForRefinement = new PredicateFactoryRefinement(services,
+				rankVarConstructor.getCsToolkitWithRankVariables().getManagedScript(), predicateFactory, false,
+				Collections.emptySet());
+		final var provider = createAutomataAbstractionProvider(services, predicateFactory, stateFactoryForRefinement,
+				witnessAutomaton);
+		final var initialAbstraction = constructInitialAbstraction(provider, icfg);
+		return new BuchiAutomatonCegarLoop<>(icfg, rankVarConstructor, predicateFactory, taPrefs, services,
+				mTransitionClazz, initialAbstraction, stateFactoryForRefinement, mCegarLoopBenchmark);
+	}
+
+	private static Set<IcfgLocation> getAcceptingStates(final IIcfg<?> icfg) {
+		final Set<IcfgLocation> allStates =
+				icfg.getProgramPoints().values().stream().flatMap(x -> x.values().stream()).collect(Collectors.toSet());
+		if (LTLPropertyCheck.getAnnotation(icfg) == null) {
+			return allStates;
+		}
+		return allStates.stream().filter(a -> BuchiProgramAcceptingStateAnnotation.getAnnotation(a) != null)
+				.collect(Collectors.toSet());
+	}
+
+	private IInitialAbstractionProvider<L, ? extends INestedWordAutomaton<L, IPredicate>>
+			createAutomataAbstractionProvider(final IUltimateServiceProvider services,
+					final PredicateFactory predicateFactory, final PredicateFactoryRefinement stateFactory,
+					final INwaOutgoingLetterAndTransitionProvider<WitnessEdge, WitnessNode> witnessAutomaton) {
+		final IInitialAbstractionProvider<L, INestedWordAutomaton<L, IPredicate>> provider =
+				new NwaInitialAbstractionProvider<>(services, stateFactory, true, predicateFactory);
+		if (witnessAutomaton == null) {
+			return provider;
+		}
+		return new WitnessAutomatonAbstractionProvider<>(services, predicateFactory, stateFactory, provider,
+				witnessAutomaton, Property.TERMINATION);
```

Comment:
**Question:** If I'm reading it correctly, the code in `BuchiCegarLoop` modified the witness automaton to make all states accepting. Is this not necessary? Or does it happen somewhere else now?

Reply:
You are right, I just fixed that.